lib/ansible/modules/clustering/kubevirt/k8s_service.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2018, KubeVirt Team <@kubevirt>
+# Apache License, Version 2.0
+# (see LICENSE or http://www.apache.org/licenses/LICENSE-2.0)
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: kubevirt_service
+
+short_description: Manage Services on Kubernetes
+
+description:
+    - Use Openshift Python SDK to manage Services on Kubernetes
+
+version_added: "2.8"
+
+author: KubeVirt Team (@kubevirt)
+
+options:
+    state:
+        description:
+            - Determines if an object should be created, patched, or deleted. When set to C(present), an object will be created, if it does not already exist. If set to C(absent), an existing object will be deleted. If set to C(present), an existing object will be patched, if its attributes differ from those specified.
+        default: present
+        choices:
+            - present
+            - absent
+        type: str
+    name:
+        description:
+            - Use to specify a Service object name.
+        required: true
+        type: str
+    namespace:
+        description:
+            - Use to specify a Service object namespace.
+        required: true
+        type: str
+    selector:
+        description:
+            - TODO
+        default: []
+        type: dict
+    type:
+        description:
+            - TODO
+        default: ClusterIP
+        choices:
+            - NodePort
+            - ClusterIP
+            - LoadBalancer
+    ports:
+        description:
+            - TODO
+        required: true
+        type: list
+    labels:
+        description:
+            - Labels are key/value pairs that are attached to objects.
+        default: {}
+        type: dict
+    spec:
```

Comment:
I think it's called ` resource_definition ` and aliases definition, inline.

Reply:
They're not the same thing. `resource_definition` in upstream k8s encompasses the whole object passed to the cluster. In my code `spec` only extends/overrides the `spec:` part of a Service definition. Naming it anything other than `spec` would imply that it does something other than what it actually does.

Reply:
Ok, sounds reasonable, but in that case we need to always support everything which isn't part of spec.

Reply:
That's an interesting problem. The solutions are:
1. Add `metadata:`, analogous to `spec`, to circumvent the issue.
2. Go through the v1 api and ensure I support all the metadata fields in existence.

Though if I go do the latter, I might as well add all the `spec` fields in existence as module options and drop `spec:` entirely. It did cross my mind, though I suspect that might end up a mine field.

Yup: I've added `labels`, but forgot for example `annotations`. I think it might make sense to just drop `labels:` and go with the `metadata:` approach. Then we're covered.

Though to be safe I'll try to dig up a full schema for Services, brb.

Reply:
I think full schema is better as with your current solution, you would need to maintain new parameters still watch if there is any new etc. With full schema you just don't care.

Reply:
Yeah, I need to add `metadata:` (and drop `labels:`), since there's [just too much of it](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#objectmeta-v1-meta) plus there's likely to be more with newer k8s releases. I should probably also add `metadata:` to `kubevirt_vm`.

`spec:` is also [rather large](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#servicespec-v1-core), which makes me think it's pointless to try to map all the arguments and instead the goal should be to support the required/most often used arguments and leave everything else in `spec:`.

Reply:
No need to add it to `kubevirt_vm` `metadata` is supported by `resource_definition` already.

Reply:
I've just pushed a version that drops this entirely and relies on `definition:` for any additional metadata (like labels).